contracts/epochs/FlowEpoch.cdc

```diff
@@ -119,36 +131,458 @@ pub contract FlowEpochs {
             self.endView = endView
             self.collectorClusters = collectorClusters
             self.clusterQCs = clusterQCs
-            self.dkgGroupKey = dkgGroupKey
+            self.dkgKeys = dkgKeys
+        }
+
+        pub fun setClusterQCs(qcs: [String]) {
+            self.clusterQCs = qcs
+        }
+
+        pub fun setDKGGroupKey(keys: [String]) {
+            self.dkgKeys = keys
+        }
+    }
+
+    /// The counter, or ID, of the current epoch
+    pub var currentEpochCounter: UInt64
+
+    /// The current phase that the epoch is in
+    pub var currentEpochPhase: EpochPhase
+
+    /// The number of views in an entire epoch
+    pub var numViewsInEpoch: UInt64
+
+    /// The number of views in the staking auction
+    pub var numViewsInStakingAuction: UInt64
+    
+    /// The number of views in each dkg phase
+    pub var numViewsInDKGPhase: UInt64
+
+    /// The number of collector clusters in each epoch
+    pub var numCollectorClusters: UInt16
+
+    /// Contains a historical record of the metadata from all previous epochs
+    /// indexed by epoch number
+    access(contract) var epochMetadata: {UInt64: EpochMetadata}
+
+    access(contract) let QCAdmin: @FlowEpochClusterQC.Admin
+
+    access(contract) let DKGAdmin: @FlowDKG.Admin
+
+    access(contract) let stakingAdmin: @FlowIDTableStaking.Admin
+
+    /// Resource that can update certain some of the contract fields
+    pub resource Admin {
+        pub fun updateEpochLength(newEpochViews: UInt64) {
+            if FlowEpoch.currentEpochPhase != EpochPhase.STAKINGAUCTION {
+                return
+            }
+
+            FlowEpoch.numViewsInEpoch = newEpochViews
+        }
+
+        pub fun updateAuctionLength(newAuctionViews: UInt64) {
+            if FlowEpoch.currentEpochPhase != EpochPhase.STAKINGAUCTION {
+                return
+            }
+
+            FlowEpoch.numViewsInStakingAuction = newAuctionViews
+        }
+
+        pub fun updateDKGPhaseLength(newPhaseViews: UInt64) {
+            if FlowEpoch.currentEpochPhase != EpochPhase.STAKINGAUCTION {
+                return
+            }
+
+            FlowEpoch.numViewsInDKGPhase = newPhaseViews
+        }
+
+        pub fun updateNumCollectorClusters(newNumClusters: UInt16) {
+            if FlowEpoch.currentEpochPhase != EpochPhase.STAKINGAUCTION {
+                return
+            }
+
+            FlowEpoch.numCollectorClusters = newNumClusters
+        }
+    }
+
+    pub resource Heartbeat {
+        /// Function that is called every block to advance the epoch
+        pub fun advanceBlock(randomSource: String?, ) {
+
+            let currentBlock = getCurrentBlock()
+            let currentEpochMetadata = FlowEpoch.epochMetadata[FlowEpoch.currentEpochCounter]!
+
+            if FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION {
+                let stakingAuctionFinalView = currentEpochMetadata.startView + FlowEpoch.numViewsInStakingAuction - 1 as UInt64
```

Comment:
What do you mean by a virtual field? Should it just be a field in `currentEpochMetadata`? Or are you saying it should be inferred? I kind of like having it as a clear enum so that it is easily queryable, but I understand that it might not be totally necessary

Reply:
Or do you mean `stakingAuctionFinalView`?

Reply:
I was referring to the `stakingAuctionFinalView`. Maybe "virtual field" isn't the right term -- I mean making `stakingAuctionFinalView` a field/method that is computed only from fields in `EpochMetadata` and isn't dependent on mutable contract-global state.

More generally I think properties of a particular epoch should be computed from an immutable epoch-specific configuration rather than the current global configuration, since we allow the global configuration to be changed over time.

Concretely, instead of doing:
```
let stakingAuctionFinalView = currentEpochMetadata.startView + FlowEpoch.numViewsInStakingAuction - 1 as UInt64
```

we would do something like:
```
pub struct EpochMetadata {
    pub fun stakingAuctionFinalView() UInt64 {
        return self.startView + self.config.numViewsInStakingAuction - 1 as UInt64
    }
// ...
}
```

Reply:
oh yeah that makes sense. I can change it to something like that. We could just make it another field in the epoch metadata struct. Would that make more sense, so it would be immutable?

Reply:
Yeah, a literal field that's computed when we construct the `EpochMetadata` instance, or a computed property based on immutable fields of `EpochMetadata`, either one works for me 👍

Reply:
ok. I made it a immutable field of epochMetadata